internal/impl/snowflake/streaming/rest.go

```diff
@@ -0,0 +1,478 @@
+	return
+}
+
+// openChannel opens a channel for writing
+func (c *SnowflakeRestClient) openChannel(ctx context.Context, req openChannelRequest) (resp openChannelResponse, err error) {
+	requestID := uuid.NewString()
+	err = c.doPost(ctx, fmt.Sprintf("https://%s.snowflakecomputing.com/v1/streaming/channels/open?requestId=%s", c.account, requestID), req, &resp)
+	return
+}
+
+// dropChannel drops a channel when it's no longer in use.
+func (c *SnowflakeRestClient) dropChannel(ctx context.Context, req dropChannelRequest) (resp dropChannelResponse, err error) {
+	requestID := uuid.NewString()
+	err = c.doPost(ctx, fmt.Sprintf("https://%s.snowflakecomputing.com/v1/streaming/channels/drop?requestId=%s", c.account, requestID), req, &resp)
+	return
+}
+
+// registerBlob registers a blob in object storage to be ingested into Snowflake.
+func (c *SnowflakeRestClient) registerBlob(ctx context.Context, req registerBlobRequest) (resp registerBlobResponse, err error) {
+	requestID := uuid.NewString()
+	err = c.doPost(ctx, fmt.Sprintf("https://%s.snowflakecomputing.com/v1/streaming/channels/write/blobs?requestId=%s", c.account, requestID), req, &resp)
+	return
+}
+
+func (c *SnowflakeRestClient) doPost(ctx context.Context, url string, req any, resp any) error {
+	marshaller := json.Marshal
+	if debug {
+		marshaller = func(v any) ([]byte, error) {
+			return json.MarshalIndent(v, "", "  ")
+		}
+	}
+	reqBody, err := marshaller(req)
+	if err != nil {
+		return err
+	}
+	respBody, err := backoff.RetryNotifyWithData(func() ([]byte, error) {
+		if debug {
+			fmt.Printf("making request to %s with body %s\n", url, reqBody)
+		}
+		httpReq, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(reqBody))
+		if errors.Is(err, context.Canceled) {
+			return nil, backoff.Permanent(err)
+		} else if err != nil {
+			return nil, fmt.Errorf("unable to make http request: %w", err)
+		}
+		httpReq.Header.Add("Content-Type", "application/json")
+		httpReq.Header.Add("Accept", "application/json")
+		httpReq.Header.Add("User-Agent", c.userAgent)
+		jwt := c.cachedJWT.Load()
+		httpReq.Header.Add("Authorization", "Bearer "+jwt)
+		httpReq.Header.Add("X-Snowflake-Authorization-Token-Type", "KEYPAIR_JWT")
+		r, err := c.client.Do(httpReq)
+		if errors.Is(err, context.Canceled) {
+			return nil, backoff.Permanent(err)
+		} else if err != nil {
+			return nil, fmt.Errorf("unable to perform http request: %w", err)
+		}
+		respBody, err := io.ReadAll(r.Body)
+		_ = r.Body.Close()
+		if errors.Is(err, context.Canceled) {
+			return nil, backoff.Permanent(err)
+		} else if err != nil {
+			return nil, fmt.Errorf("unable to read http response: %w", err)
+		}
+		if r.StatusCode != 200 {
+			return nil, fmt.Errorf("non successful status code (%d): %s", r.StatusCode, respBody)
+		}
+		if debug {
+			fmt.Printf("got response to %s with body %s\n", url, respBody)
```

Comment:
Would be nice to dump this data in the log at trace level if someone needs to troubleshoot it.

Reply:
Yeah let me add that.